File: services/ai/utils/confidence.py
from typing import Any, Dict, Optional

from pydantic import BaseModel, Field


class ConfidenceInterval(BaseModel):
    low: float = Field(ge=0.0, le=1.0)
    high: float = Field(ge=0.0, le=1.0)


class CostFields(BaseModel):
    tokenUsage: int = 0
    computeUsage: float = 0.0
    storageUsage: float = 0.0
    estimatedCost: float = 0.0


class ResultMetadata(BaseModel):
    confidence: float = Field(ge=0.0, le=1.0)
    confidenceInterval: ConfidenceInterval
    modelVersion: str
    evaluationVersion: str
    tokenUsage: int = 0
    computeUsage: float = 0.0
    storageUsage: float = 0.0
    estimatedCost: float = 0.0


def build_result_metadata(
    confidence: float,
    model_version: str,
    evaluation_version: str = "eval-1.0.0",
    margin: float = 0.05,
    token_usage: int = 0,
    compute_usage: float = 0.0,
    storage_usage: float = 0.0,
    estimated_cost: float = 0.0,
) -> Dict[str, Any]:
    """Build standard metadata fields required on every result payload."""
    low = max(0.0, confidence - margin)
    high = min(1.0, confidence + margin)
    return {
        "confidence": round(confidence, 4),
        "confidenceInterval": {"low": round(low, 4), "high": round(high, 4)},
        "modelVersion": model_version,
        "evaluationVersion": evaluation_version,
        "tokenUsage": token_usage,
        "computeUsage": compute_usage,
        "storageUsage": storage_usage,
        "estimatedCost": estimated_cost,
    }


def merge_metadata(base: Dict[str, Any], extra: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
    result = dict(base)
    if extra:
        result.update(extra)
    return result

File: apps/backend/src/modules/ai/utils/confidence.ts
export type ConfidenceBundle = {
  confidence: number;
  confidenceInterval: { low: number; high: number };
  confidenceLow: number;
  confidenceHigh: number;
  modelVersion: string;
  evaluationVersion: string;
};

export type CostBundle = {
  tokenUsage: number;
  computeUsage: number;
  storageUsage: number;
  estimatedCost: number;
};

export function buildConfidence(input?: {
  confidence?: number;
  low?: number;
  high?: number;
  modelVersion?: string;
  evaluationVersion?: string;
}): ConfidenceBundle {
  const confidence = clamp01(input?.confidence ?? 0.72);
  const low = clamp01(input?.low ?? Math.max(0, confidence - 0.12));
  const high = clamp01(input?.high ?? Math.min(1, confidence + 0.08));
  return {
    confidence,
    confidenceInterval: { low, high },
    confidenceLow: low,
    confidenceHigh: high,
    modelVersion: input?.modelVersion ?? "MODEL-VERSION-LOCAL001",
    evaluationVersion: input?.evaluationVersion ?? "eval-1.0.0",
  };
}

export function buildCost(input?: Partial<CostBundle>): CostBundle {
  return {
    tokenUsage: input?.tokenUsage ?? 0,
    computeUsage: input?.computeUsage ?? 5,
    storageUsage: input?.storageUsage ?? 0,
    estimatedCost: input?.estimatedCost ?? 0,
  };
}

function clamp01(n: number): number {
  if (Number.isNaN(n)) return 0;
  return Math.min(1, Math.max(0, n));
}
